ScaffoldGeometry plumbing.

amirh opened the pull request:
Adds a ScaffoldGeometry class and ValueNotifier for it.
A scaffold's ScaffoldGeometry notifier is held in the _ScaffoldState,
and is passed to _ScaffoldScope.
New ScaffoldGemometry objects are built and published to the notifier
during the layout pass after computing the scaffold's layout.

amirh committed:
ScaffoldGeometry plumbing.

Adds a ScaffoldGeometry class and ValueNotifier for it.
A scaffold's ScaffoldGeometry notifier is held in the _ScaffoldState,
and is passed to _ScaffoldScope.
New ScaffoldGemometry objects are built and published to the notifier
during the layout pass after computing the scaffold's layout.

amirh commented:
@Hixie @DaveShuckerow @HansMuller here's a quick suggestion of how to plumb the ScaffoldGeometry change notifications.

Not tests yet, as I'm first seeking your feedback on the way it's plumbed.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold gemoetry call
```
(typo in geometry)

HansMuller replied:
In the fullness of time, this doc should explain what "geometry" means in this context, why you'd want to listen to this notifier, and when you'd want to handle notifications.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
+  /// Null when there is no bottom navigation bar..
+  final Rect floatingActionButton;
+
+  const ScaffoldGeometry({
```
nit: constructors first (when first reading this i assumed there was an implicit constructor, but when i got to this constructor i realised no, actually, there's no implicit constructor)

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.bottomNavigationBar].
+  /// Null when there is no bottom navigation bar..
+  final Rect bottomNavigationBar;
```
This could probably just be a double giving the distance from some reference edge to the top of the navbar area. That would reduce the temptation to use this value to determine geometry that is best discovered by directly examining the navbar.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  /// Null when there is no bottom navigation bar..
+  final Rect floatingActionButton;
```
similarly here i'd probably express this in terms of the offset from some reference point to the center of the fab, and a double giving the current fractional size of the fab. That would give more flexibility, e.g. in case the shape of the fab is not well described by a Rect (e.g. it's a T-shape, or a diamond).

amirh replied:
This will be a fraction of what? of some fixed fab size? if we want to reduce this to a single double maybe we can use the size of one of the edges, or the diagonal, though I'm not sure this the best thing to do here...

Your point for reducing the temptation to use the bottomNavigationBarRect makes sense when thinking of the BAB use case, where we want to at least look at the BAB's size.

For the FAB dimensions though, there is no such temptation, in the BAB use case all you know about the FAB comes from ScaffoldGeometry.

I would leave it a Rect as it contains the full information of where the scaffold is laying out the FAB in a simple form. Also you could think of a case where the width/height proportions are changing, and a fractional size is not enough.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
+  /// [Scaffold] ancestor of the given context.
+  ///
+  /// Value notifications are sent during the layout pass.
+  static ValueNotifier<ScaffoldGeometry> geometryOf(BuildContext context) {
```
to avoid anyone being tempted to set the value directly on this object, maybe make this return a ValueListenable instead of a ValueNotifier. (ValueNotifier implements ValueListenable so no code need change.)

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
+  /// Returns a [ValueNotifier] for the [ScaffoldGeometry] for the closest
+  /// [Scaffold] ancestor of the given context.
+  ///
+  /// Value notifications are sent during the layout pass.
```
docs should talk about how this object must be discarded any time the given context has its dependencies change, notably, for the context from a StatefulWidget, whenever the corresponding State's didChangeDepndencies is fired.

amirh replied:
Added some details to the doc

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
+  static ValueNotifier<ScaffoldGeometry> geometryOf(BuildContext context) {
+    final _ScaffoldScope scaffoldScope = context.inheritFromWidgetOfExactType(_ScaffoldScope);
+    if (scaffoldScope == null)
+    throw new FlutterError(
```
nits: indent, braces

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
+      'your new inner widgets, and then in these inner widgets you would use Scaffold.geometryOf().\n'
+      'A less elegant but more expedient solution is assign a GlobalKey to the Scaffold, '
+      'then use the key.currentState property to obtain the ScaffoldState rather than '
+      'using the Scaffold.geometryOf() function.\n'
```
this message doesn't make sense. You can't use a GlobalKey to get the geometry value notifier.

amirh replied:
oops, clumsy copy-pasting

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
+class ScaffoldGeometry {
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.bottomNavigationBar].
+  /// Null when there is no bottom navigation bar..
```
nit: these docs need work ("Null" should be lowercase, since Null is the type and null is the value; double period at the end; blank line should be between the two sentences, needs more detail)

amirh replied:
change the nits, though I'm not sure which more details are missing here.
(though we might be changing this to not be Rect at all, so maybe not relevant)

Hixie commented:
I recommended some trivial tweaks, but the general approach seems solid to me.

One thing I would look into is how the fab shrink/grow logic is implemented. Somehow that needs to be exposed in this geometry object; it's possible we implement that efficiently enough that it's entirely done as an animation/paint loop, no layout, in which case we presumably want to somehow get the relevant Animation involved here.

amirh commented:
Good point regarding the FAB scaling logic, it is indeed implemented with a Transform that's skipping layout. I'm make it update geometry as well, though it makes its contract a bit more complex (we can no longer say it's updated as part of the layout pass)

amirh committed:
review comments followup, and add tests

amirh commented:
PTAL, added tests, and added geometry notifications from the FAB transition animation.

TBD in following CLs - add the 'notchMaker' closure we discussed to ScaffoldGeometry

DaveShuckerow commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
+class ScaffoldGeometry {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButton,
```
I think we're going to need two different types of ScaffoldGeometry.  The fab placement logic I'm working on depends on the scaffold's geometry for figuring out what offset to give the fab, so the position rect won't apply.

I think the easiest thing to do will be to pass the fab positioner an incomplete scaffold geometry, then we can finish populating it with the fab layout after it's computed.  

### TLDR
So long as it's possible to build ScaffoldGeometry without data about the fab's layout and without passing the incomplete object into the value notifier, this shouldn't be a problem.

amirh replied:
Will we be able to populate the floatingActionButton with it's position before paint time?

As long as by paint time we populated the fab information in the geometry object it should be ok.
In current CL we already populate it with incomplete data from the FabTransition before the first layout.

DaveShuckerow replied:
Yes.  The way it works in #14368 currently is:

1. Scaffold passes FabPositioner and FabMotionAnimator to the Layout delegate
2. Layout delegate computes layout of everything but the fab and creates a ScaffoldGeometry
3. FabPositioner uses the ScaffoldGeometry to determine the Fab's Offset
4. Layout delegate computes the Fab's layout based on the requested offset (interpolated by animation when necessary)

The output of this can be a ScaffoldGeometry that includes the Fab's layout rect, and that will be accessible as part of the Layout delegate's layout phase, before painting.

amirh replied:
great

Hixie replied:
it seems weird to me to use the same struct for computing the fab position and for obtaining the fab position

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
+  /// set.
+  final Rect floatingActionButton;
```
is a Rect really the best way to describe this? Consider a FAB whose shape is a downwards-pointing triangle.

I'd probably just go with an offset to the center, and a radius?

Anyway, a Rect is fine if you think it's better than that.

amirh replied:
Following f2f we keep it a rect, as this keeps the notch computation logic similar to the path computation logic for the ShapeBorder (both take a Rect as their input)

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
+  final Rect floatingActionButton;
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  final double floatingActionButtonScale;
```
This seems redundant with the Rect?

Hixie replied:
At a minimum the docs should explain in detail how the scale and the position of the fab interact.

amirh replied:
Added docs

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
+  /// set.
+  final Rect floatingActionButton;
```
we should probably call this floatingActionButtonPosition or something

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  final double floatingActionButtonScale;
+
+  static ScaffoldGeometry _update({
```
this should just use the copyWith pattern

amirh replied:
renamed to _copyWith, is that what you meant?

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -284,6 +347,24 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
+  }
+
+  void _updateGeometryScale(double scale) {
+    widget.geometryNotifier.value = ScaffoldGeometry._update(
```
this is going to cause the notifier to notify multiple times per layout, which seems suboptimal.

Hixie replied:
we discussed this and couldn't find a workaround.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -514,6 +595,50 @@ class Scaffold extends StatefulWidget {
+  /// the listener, and register a listener to the new [ScaffoldGeometry]
+  /// listenable.
+  // TODO(amirh): return a ValueListenable subclass that asserts it's paint time
+  // when .value is accessed.
```
we should just do that now, otherwise we'll never be able to do it because people will depend on us not doing it

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -798,6 +925,12 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+  @override
+  void initState() {
+    super.initState();
+    _geometryNotifier = new ValueNotifier<ScaffoldGeometry>(null);
```
dispose of this in the dispose method

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -212,15 +270,20 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
   @override
   void dispose() {
+    _previousAnimation.removeListener(_onProgressChanged);
+    _currentAnimation.removeListener(_onProgressChanged);
```
you don't need to remove these, the dispose methods below will do it for you

amirh replied:
done

amirh committed:
review comments followup

amirh commented:
PTAL

amirh committed:
remove blank line

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButtonPosition,
+    this.floatingActionButtonScale = 1.0,
```
:

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  ///
+  /// The floating action button might be scaled inside this rectangle, to get
+  /// the bounding rectangle in which the floating action is painted scale this
+  /// value by [floatingActionButtonScale].
```
It's not clear what "scale" means here. Maybe show same `## Sample code` for how to do that?

amirh replied:
done.
@Hixie looking at it again I don't like the Position suffix we added to floatingActionButton, as it sounds like a point and not a rectangle. What do you think about just calling it floatingActionButton (geometry.floatingActionButton sounds pretty reasonable I believe).
If not, maybe floatingActionButtonRect ?

Hixie replied:
A "floatingActionButton" property should hold a floating action button.
I agree that "position" suggests an Offset.

This gives you the area that the FAB would cover if it was not scaled, right?
Maybe "floatingActionButtonArea"?

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+
+  @override
+  ScaffoldGeometry get value{
+    if (!RendererBinding.instance.pipelineOwner.debugDoingPaint)
```
that might not be the right PipelineOwner.

You want the owner for the RenderObject that's actually doing the scaffold layout.

amirh replied:
done.
I'm not sure how expensive is the findRenderObject, should/can I cache the pipeline owner once I find it?

Hixie replied:
Just put all this code in an assert.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  ScaffoldGeometry get value{
+    if (!RendererBinding.instance.pipelineOwner.debugDoingPaint)
+      throw new FlutterError(
+        'Scaffold.geometryOf() value was accessed not during the paint phase.'
```
we should phrase this as something like "Scaffold.geometryOf() value must only be accessed during the paint phase.", otherwise the logical reaction to that error message is "I know, so what".

Hixie replied:
Maybe even make this a multiline message:

```
Scaffold.geometryOf() value obtained outside paint phase.
The ScaffoldGeometry is only available during the paint phase, because its value is computed during
the animation and layout phases prior to painting. If you are trying to obtain the value during the
build or layout phases, consider moving your logic to the paint phase, since the value is not yet
completely available during build or layout. If you are trying to obtain the value between frames,
consider caching the value during paint.
```

...or something

amirh replied:
I sub-sampled parts of your message :)

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -212,11 +294,14 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
     // of animating in.
     if (widget.child != null)
       _currentController.value = 1.0;
+    else
```
there's a comment above explaining the if statement but its explanation is now incomplete

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -548,6 +693,8 @@ class Scaffold extends StatefulWidget {
 /// the current [BuildContext] using [Scaffold.of].
 class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
 
+  _ScaffoldGeometryNotifier _geometryNotifier;
```
we should put this in the section of this class that uses it, i.e. below the `// INTERNALS` comment above initState

amirh replied:
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  /// the bounding rectangle in which the floating action is painted scale this
+  /// value by [floatingActionButtonScale].
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
```
For the BAB needs it is not a problem: when there is no FAB the scale is 0, so the BAB won't create a notch.

Setting this to null when there's no FAB makes the dance between the transition animation ticks and the layout code even more complex (when you set the floatingActionButton to null, there is still a FAB until the transition animation is complete). Which I rather avoid at this stage (we expect some more churn to this code with Dave's work, once it stabilize we can see if it's reasonable).

I added to the doc comment that when there is no FAB the scale will be 0.

Hixie replied:
I guarantee you that eventually someone will write code that depends on exactly what value you export here when the scale is 0. If we're going to expose a non-null value there in that case, I think we should at a minimum document exactly what the value will be, even if it isn't very useful. The reality is that eventually (maybe in a few years) we'll be forced to be compatible with whatever value we expose here even if we want to change it.

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  ///
+  /// To et the bounding rectangle in shich the floating action button is
```
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  ///
+  /// To et the bounding rectangle in shich the floating action button is
+  /// painted scale [floatingActionPosition] by this proportion. 
```
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+  /// painted scale [floatingActionPosition] by this proportion. 
+  final double floatingActionButtonScale;
+
+  static ScaffoldGeometry _copyWith({
```
@Hixie  I'm sorry but I do not know what do you mean by the "copywith pattern", can you be more explicit? do you mean you want the method non static? it won't work in this case as previousGeometry might be null.

amirh replied:
I moved the partial update logic into _ScaffoldGeometryNotifier, this way the partial value is not accessible even from the scaffold code.

Still curious to know what do you mean by the "copywith pattern"

Hixie replied:
If you look around for `copyWith` methods you'll see what I mean. Lots of our classes have a method named `copyWith` that allows the class to be copied with edits.

To solve the "null" problem, just initialize the ValueNotifier with a non-null value.

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
+    : super(geometry);
+
+  @override
+  ScaffoldGeometry get value{
```
done

amirh committed:
review comments followup

amirh commented:
PTAL

amirh committed:
move the gemoetry partial logic update into the notifier

amirh committed:
review comments followup

amirh commented:
PTAL

amirh committed:
fix sample code

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
+      final RenderObject renderObject = context.findRenderObject();
+      if (renderObject == null || !renderObject.owner.debugDoingPaint)
+        throw new FlutterError(
+            'Scaffold.geometryOf() must only be accessed during the paint phase.\n\n'
```
only one `\n` after the first line (see also `FlutterError.message` documentation)

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
+        throw new FlutterError(
+            'Scaffold.geometryOf() must only be accessed during the paint phase.\n\n'
+                'The ScaffoldGeometry is only available during the paint phase, because\n'
+                'its value is computed during the animation and layout phases prior to painting.'
```
inconsistent indentation

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
+    double floatingActionButtonScale,
+  }) {
+
+    final double newFloatingActionButtonScale = floatingActionButtonScale ??  super.value?.floatingActionButtonScale;
```
nit: extra spaces here and below after `??`

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
+    Rect floatingActionButtonArea,
+    double floatingActionButtonScale,
+  }) {
+
```
nit: extraneous blank line

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

amirh committed:
review comments followup
